step_shutdown.go

```diff
@@ -41,38 +44,71 @@ func (s *StepShutdown) Run(state multistep.StateBag) multistep.StepAction {
 			return multistep.ActionHalt
 		}
 
-		// TODO: add timeout
-		for !cmd.Exited {
-			ui.Say("Waiting for remote cmd to finish...")
-			time.Sleep(150 * time.Millisecond)
-		}
-		if cmd.ExitStatus != 0 && cmd.ExitStatus != packer.CmdDisconnect {
-			err := fmt.Errorf("Cmd exit status %v, not 0", cmd.ExitStatus)
-			state.Put("error", err)
-			ui.Error(err.Error())
-			return multistep.ActionHalt
-		} else if cmd.ExitStatus == packer.CmdDisconnect {
-			ui.Say("VM disconnected")
+		// Wait for the machine to actually shut down
+		log.Printf("Waiting max %s for shutdown to complete", s.ShutdownTimeout)
+		shutdownTimer := time.After(s.ShutdownTimeout)
+		for {
+			powerState, err := vm.PowerState(ctx)
+			if err != nil {
+				state.Put("error", err)
+				return multistep.ActionHalt
+			}
+			if powerState == "poweredOff" {
+				break
+			}
+
+			select {
+			case <-shutdownTimer:
+				log.Printf("Shutdown stdout: %s", stdout.String())
+				log.Printf("Shutdown stderr: %s", stderr.String())
+				err := errors.New("Timeout while waiting for machine to shut down.")
+				state.Put("error", err)
+				ui.Error(err.Error())
+				return multistep.ActionHalt
+			default:
+				time.Sleep(150 * time.Millisecond)
+			}
 		}
 	} else {
 		ui.Say("Forcibly halting virtual machine...")
 
 		err := vm.ShutdownGuest(ctx)
 		if err != nil {
-			state.Put("error", err)
+			state.Put("error", fmt.Errorf("Could not shutdown guest: %v", err))
 			return multistep.ActionHalt
 		}
 
-		task, err := vm.PowerOff(ctx)
-		if err != nil {
-			state.Put("error", err)
-			return multistep.ActionHalt
+		// Wait for the guest to actually shut down
+		log.Printf("Waiting max %s for guest shutdown to complete", s.ShutdownTimeout)
+		shutdownTimer := time.After(s.ShutdownTimeout)
+		var vmImage mo.VirtualMachine
+		for {
+			err = vm.Properties(ctx, vm.Reference(), []string{"guest.guestState"}, &vmImage)
```

Comment:
As I understand, they are just two different entities. But I may be wrong because, indeed,  after shuttung down the guest system power off is impossible.